File: jonty-processor/src/main/java/com/petertackage/jonty/compiler/JontyProcessor.java
/*
 * Copyright 2017 Futurice GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.petertackage.jonty.compiler;

import com.google.auto.service.AutoService;
import com.petertackage.jonty.Fieldable;
import com.squareup.kotlinpoet.ClassName;
import com.squareup.kotlinpoet.FileSpec;

import javax.annotation.processing.*;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.Elements;
import javax.tools.Diagnostic;
import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.lang.annotation.Annotation;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

@AutoService(Processor.class)
public final class JontyProcessor extends AbstractProcessor {

    private static final String OPTION_DEBUGGABLE = "jonty.debuggable";

    private Elements elementUtils;
    private Filer filer;
    private boolean debuggable = true;

    public JontyProcessor() {
        super();
    }

    @Override
    public synchronized void init(ProcessingEnvironment env) {
        super.init(env);
        elementUtils = env.getElementUtils();
        filer = env.getFiler();
        debuggable = !"false".equals(env.getOptions().get(OPTION_DEBUGGABLE));
    }

    @Override
    public boolean process(final Set<? extends TypeElement> annotations,
                           final RoundEnvironment roundEnv) {

        note(null, "Starting processings");

        Map<TypeElement, Fielder> fielderMap = findAndParseFields(roundEnv);

        for (Map.Entry<TypeElement, Fielder> entry : fielderMap.entrySet()) {
            TypeElement typeElement = entry.getKey();
            Fielder fielder = entry.getValue();

            // Write it out to a file.
            FileSpec kotlinFile = fielder.brew(debuggable);

            try {
                kotlinFile.writeTo(new File("/Users/ptac/gen"));
            } catch (IOException e) {
                error(typeElement, "Unable to write fielder for type %s: %s", typeElement, e.getMessage());
            }

        }

        return false;
    }

    @Override
    public SourceVersion getSupportedSourceVersion() {
        return SourceVersion.latestSupported();
    }

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        return Collections.singleton(Fieldable.class.getCanonicalName());
    }

    private Map<TypeElement, Fielder> findAndParseFields(RoundEnvironment env) {

        // Maps of classes to builders
        Map<TypeElement, Fielder.Builder> builderMap = new LinkedHashMap<>();

        // Process each @Fieldable class element.
        for (Element element : env.getElementsAnnotatedWith(Fieldable.class)) {
            note(element, "Got this element %s", element);
            try {
                parseFieldableClass(element, builderMap);
            } catch (Exception e) {
                logParsingError(element, Fieldable.class, e);
            }
        }

        // @TODO Traverse up the tree, but without duplicates. Don't need to do this yet.
        // @TODO Will just assume that only one annotation exists in the hierachy.
        // @TODO In any case, duplicates aren't problematic
        // Associate superclass binders with their subclass binders. This is a queue-based tree walk
        // which starts at the roots (superclasses) and walks to the leafs (subclasses).
        Map<TypeElement, Fielder> fielderMap = new LinkedHashMap<>();

        for (Map.Entry<TypeElement, Fielder.Builder> entry : builderMap.entrySet()) {
            fielderMap.put(entry.getKey(), entry.getValue().build());
        }
        return fielderMap;
    }

    private void parseFieldableClass(Element element,
                                     Map<TypeElement, Fielder.Builder> builderMap) {

        TypeElement typeElement = (TypeElement) element;
        String packageName = getPackageName(typeElement);
        String className = getClassName(typeElement, packageName);
        ClassName bindingClassName = new ClassName(packageName, className + "_Fielder");

        Fielder.Builder fielderBuilder = new Fielder.Builder(bindingClassName);

        for (Element enclosedElement : element.getEnclosedElements()) {

            // Only interested in fields
            if (enclosedElement.getKind() == ElementKind.FIELD) {
                String fieldName = enclosedElement.getSimpleName().toString();
                fielderBuilder.addName(fieldName);
            }

        }

        builderMap.put((TypeElement) element, fielderBuilder);
    }

    private static String getClassName(TypeElement type, String packageName) {
        return packageName.length() == 0 ?
                type.getQualifiedName().toString() :
                type.getQualifiedName().toString().substring(packageName.length() + 1)
                        .replace('.', '$');
    }

    private String getPackageName(TypeElement type) {
        return elementUtils.getPackageOf(type).getQualifiedName().toString();
    }

    private void logParsingError(Element element, Class<? extends Annotation> annotation,
                                 Exception e) {
        StringWriter stackTrace = new StringWriter();
        e.printStackTrace(new PrintWriter(stackTrace));
        error(element, "Unable to parse @%s fielder.\n\n%s", annotation.getSimpleName(),
                stackTrace);
    }

    private void error(Element element, String message, Object... args) {
        printMessage(Diagnostic.Kind.ERROR, element, message, args);
    }

    private void note(Element element, String message, Object... args) {
        printMessage(Diagnostic.Kind.NOTE, element, message, args);
    }

    private void printMessage(Diagnostic.Kind kind, Element element, String message,
                              Object[] args) {
        if (args.length > 0) {
            message = String.format(message, args);
        }

        processingEnv.getMessager().printMessage(kind, message, element);
    }

}

File: jonty-processor/src/main/java/com/petertackage/jonty/processor/JontyProcessor.kt
/*
 * Copyright 2017 Peter Tackage
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.petertackage.jonty.processor

import com.petertackage.jonty.Fieldable
import com.petertackage.jonty.processor.internal.Fielder
import com.squareup.kotlinpoet.ClassName
import java.io.IOException
import java.io.PrintWriter
import java.io.StringWriter
import java.nio.file.Paths
import java.util.*
import javax.annotation.processing.AbstractProcessor
import javax.annotation.processing.ProcessingEnvironment
import javax.annotation.processing.RoundEnvironment
import javax.annotation.processing.SupportedOptions
import javax.lang.model.SourceVersion
import javax.lang.model.element.Element
import javax.lang.model.element.ElementKind
import javax.lang.model.element.Modifier
import javax.lang.model.element.TypeElement
import javax.lang.model.type.DeclaredType
import javax.lang.model.type.TypeKind
import javax.lang.model.type.TypeMirror
import javax.lang.model.util.Elements
import javax.tools.Diagnostic

@SupportedOptions(JontyProcessor.KAPT_KOTLIN_GENERATED_OPTION_NAME)
class JontyProcessor : AbstractProcessor() {

    private var elementUtils: Elements? = null
    private var debuggable = true

    companion object {
        const val KAPT_KOTLIN_GENERATED_OPTION_NAME = "kapt.kotlin.generated"
        private val OPTION_DEBUGGABLE = "jonty.debuggable"
        private val DEBUG = false
    }

    @Synchronized override fun init(env: ProcessingEnvironment) {
        super.init(env)
        elementUtils = env.elementUtils
        debuggable = "false" != env.options[OPTION_DEBUGGABLE]
    }

    override fun process(annotations: Set<TypeElement>,
                         roundEnv: RoundEnvironment): Boolean {

        debug(null, "Starting annotation processing round.")
        val fielderMap = findAndParseFields(roundEnv)

        for ((typeElement, fielder) in fielderMap) {

            val kotlinFile = fielder.brew()

            try {
                val kaptKotlinGeneratedOutDir = processingEnv.options[KAPT_KOTLIN_GENERATED_OPTION_NAME]
                debug(null, "kaptKotlin output directory: %s.", KAPT_KOTLIN_GENERATED_OPTION_NAME)
                kotlinFile.writeTo(Paths.get(kaptKotlinGeneratedOutDir))
            } catch (e: IOException) {
                error(typeElement, "Unable to write fielder for type %s: %s.", typeElement, e)
            }

        }

        return false
    }

    override fun getSupportedSourceVersion(): SourceVersion {
        return SourceVersion.latestSupported()
    }

    override fun getSupportedAnnotationTypes(): Set<String> {
        return setOf(Fieldable::class.java.canonicalName)
    }

    private fun findAndParseFields(env: RoundEnvironment): Map<TypeElement, Fielder> {

        // Maps of classes to builders
        val builderMap = LinkedHashMap<TypeElement, Fielder.Builder>()

        // Process each @Fieldable class element.
        for (element in env.getElementsAnnotatedWith(Fieldable::class.java)) {
            debug(element, "Processing annotated element: %s.", element)
            try {
                parseFieldableClass(element, builderMap)
            } catch (e: Exception) {
                logParsingError(element, Fieldable::class.java, e)
            }

        }

        val fielderMap = LinkedHashMap<TypeElement, Fielder>()

        for ((key, value) in builderMap) {
            fielderMap.put(key, value.build())
        }
        return fielderMap
    }

    private fun parseFieldableClass(element: Element,
                                    builderMap: MutableMap<TypeElement, Fielder.Builder>) {

        var typeElement: TypeElement? = element as TypeElement
        val packageName = getPackageName(typeElement)
        val className = getClassName(typeElement, packageName)
        val bindingClassName = ClassName(packageName, className + "_JontyFielder")

        val fielderBuilder = Fielder.Builder(bindingClassName)
        builderMap.put(element, fielderBuilder)

        while (typeElement != null) {
            for (enclosedElement in typeElement.enclosedElements) {

                debug(enclosedElement,
                        "Element enclosed element: %s, modifiers: %s, type: %s.",
                        enclosedElement.simpleName, enclosedElement.modifiers, enclosedElement.asType())

                // Only interested in non-static fields; properties
                if (enclosedElement.kind == ElementKind.FIELD && !enclosedElement.modifiers.contains(Modifier.STATIC)) {
                    val fieldName = enclosedElement.simpleName.toString()
                    debug(enclosedElement, "!! Adding field !!: %s.", fieldName)
                    fielderBuilder.addName(fieldName)
                }
            }
            // Add fields from the parent class
            typeElement = findParentType(typeElement)
        }

    }

    private fun findParentType(typeElement: TypeElement): TypeElement? {
        debug(typeElement, "Attempting to find superclass for %s.", typeElement)
        val type: TypeMirror = typeElement.superclass
        if (type.kind == TypeKind.NONE) {
            debug(typeElement, "No superclass for: %s.", type)
            return null
        }
        debug(typeElement, "Found superclass: %s.", type)
        return (type as DeclaredType).asElement() as TypeElement
    }

    private fun getPackageName(type: TypeElement?): String {
        return elementUtils!!.getPackageOf(type).qualifiedName.toString()
    }

    private fun logParsingError(element: Element, annotation: Class<out Annotation>, e: Exception) {
        val stackTrace = StringWriter()
        e.printStackTrace(PrintWriter(stackTrace))
        error(element, "Unable to parse @%s fielder.\n\n%s.", annotation.simpleName, stackTrace)
    }

    private fun error(element: Element, message: String, vararg args: Any) {
        printMessage(Diagnostic.Kind.ERROR, element, message, *args)
    }

    private fun note(element: Element?, message: String, vararg args: Any) {
        printMessage(Diagnostic.Kind.NOTE, element, message, *args)
    }

    private fun debug(element: Element?, message: String, vararg args: Any) {
        if (DEBUG) printMessage(Diagnostic.Kind.OTHER, element, message, *args)
    }

    private fun printMessage(kind: Diagnostic.Kind, element: Element?, message: String, vararg args: Any) {
        var formattedMsg = message;
        if (args.isNotEmpty()) {
            formattedMsg = message.format(*args)
        }
        processingEnv.messager.printMessage(kind, formattedMsg, element)
    }

    private fun getClassName(type: TypeElement?, packageName: String): String {
        return if (packageName.length == 0) type!!.qualifiedName.toString()
        else type!!.qualifiedName.toString().substring(packageName.length + 1)
                .replace('.', '_')
    }

}
